modules/types.ts

```diff
@@ -7,14 +7,15 @@ export interface SetStateProp<P, S> {
     setState: SetState<P, S>
 }
 
-export type MapToProps<P, S, ExtraP> = (
+export type MergeProps<P, S, ExtraP> = (
     state: S,
-    setState: SetState<P, S>
-) => ExtraP
+    setStateProps: ExtraP,
+    initialProps: P
+) => P
```

Comment:
I can't be `P` unless `initialProps` is of type `Partial<P>`. Before `BaseComponent` was called with an union type `P & S & ExtraP`. Now that we open the merge as a custom operation, we also need users to provide us with their final prop type...